[scudo] Added LRU eviction policy to secondary cache.

JoshuaMBa opened the pull request:
The logic for emptying the cache now follows an LRU eviction policy. When the cache is full on any given free operation, the oldest entry in the cache is evicted, and the memory associated with that cache entry is unmapped.

Finding empty cache entries is now a constant operation with the use of a stack of available cache entries.

Through the LRU structure, the cache retrieval algorithm now only iterates through valid entries of the cache. Furthermore, the retrieval algorithm will first search cache entries that have not been decommitted (i.e. madvise() has not been called on their corresponding memory chunks) to reduce the likelihood of returning a memory chunk to the user that would induce a page fault.

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/secondary.h:
```diff
@@ -249,11 +273,17 @@ template <typename Config> class MapAllocatorCache {
     }
     do {
       ScopedLock L(Mutex);
+
+      Time = getMonotonicTimeFast();
+      if (Entry.Time != 0)
+        Entry.Time = Time;
+
```
I know we will remove the `Entry.Time` but it's still better to have some comment on why we move the `getMonotonicTimeFast()` here.

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/secondary.h:
```diff
@@ -213,23 +225,35 @@ template <typename Config> class MapAllocatorCache {
     const u32 MaxCount = atomic_load_relaxed(&MaxEntriesCount);
+    u64 Time;
     CachedBlock Entry;
+    Vector<MemMapT, 1U> EvictionMemMaps;
```
Can we have some comment on why we use 1 as default?

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/secondary.h:
```diff
@@ -202,6 +210,10 @@ template <typename Config> class MapAllocatorCache {
+  // Ensure the cache entry array size fits in the LRU list Next and Prev
+  // index fields
+  static_assert(Config::getEntriesArraySize() <= CachedBlock::CacheIndexMax,
+                "");
```
Let's have some thing in the description. It can be a brief one and let the comment tell the detail

JoshuaMBa commented on compiler-rt/lib/scudo/standalone/secondary.h:
```diff
@@ -269,30 +299,27 @@ template <typename Config> class MapAllocatorCache {
-        }
+
+      // All excess entries are evicted from the cache
+      while (EntriesCount >= MaxCount) {
+        // Save MemMaps of evicted entries to perform unmap outside of lock
+        EvictionMemMaps.push_back(Entries[LRUTail].MemMap);
+        remove(LRUTail);
```
Just to clarify, you want `needToEvict` to decrement the `EntriesCount` variable instead? And if we do that, should something similar be done for `insert`?

ChiaHungDuan replied:
No, use it as the loop condition.
```
while (needToEvict()) {
  ...
}
```

JoshuaMBa replied:
Ok, I wonder if we should pass `MaxCount` to `needToEvict` because we don't want to perform an atomic load of `MaxEntriesCount` for every iteration, right?

ChiaHungDuan replied:
As you said, it's a rare event. In addition, it's atomic_load_relaxed so it should be a normal load. So we can do the load without performance concern

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/secondary.h:
```diff
@@ -434,6 +464,66 @@ template <typename Config> class MapAllocatorCache {
+    AvailableHead = static_cast<u16>(I);
+    EntriesCount--;
+
+    // Cache should not have valid entries when not empty
+    if (EntriesCount == 0) {
+      DCHECK(LRUHead == CachedBlock::InvalidEntry);
+      DCHECK(LRUTail == CachedBlock::InvalidEntry);
+    }
+  }
+
```
You change this to DCHECK_NE in the follow up CL, it may be better to do it here. 

I think it's fine to use DCHECK anyway because we only care if it's InvalidEntry. Knowing the wrong value may not help with figuring out what's going on. I'll leave you to make the decision

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/secondary.h:
```diff
@@ -484,14 +573,20 @@ template <typename Config> class MapAllocatorCache {
   NonZeroLengthArray<CachedBlock, Config::getQuarantineSize()>
       Quarantine GUARDED_BY(Mutex) = {};
+
+  // The LRUHead of the cache is the most recently used cache entry
+  // The LRUTail of the cache is the least recently used cache entry
+  // The AvailableHead is the top of the stack of available entries
+  u16 LRUHead GUARDED_BY(Mutex) = 0;
+  u16 LRUTail GUARDED_BY(Mutex) = 0;
+  u16 AvailableHead GUARDED_BY(Mutex) = 0;
```
I suggest inlining each comment to the associated variable in the follow up CL. Maybe we can do it here

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/secondary.h:
```diff
@@ -410,19 +442,17 @@ template <typename Config> class MapAllocatorCache {
   void disableMemoryTagging() EXCLUDES(Mutex) {
     ScopedLock L(Mutex);
-    for (u32 I = 0; I != Config::getQuarantineSize(); ++I) {
+    for (u32 I = 0; I != Config::getQuarantineSize(); I++) {
```
Revert this change

cferris1000 commented on compiler-rt/lib/scudo/standalone/secondary.h:
```diff
@@ -213,23 +225,39 @@ template <typename Config> class MapAllocatorCache {
+    // Only in the rare event that the cache shrinks in real-time
+    // due to a decrease in the configurable value MaxEntriesCount
+    // will more than one cache entry be evicted
+    Vector<MemMapT, 1U> EvictionMemMaps;
```
I'd move this right in front of the do loop down below so it's more obvious where this is used. I'd also add a comment explaining that you using this to gather the maps to evict and then doing it outside of the lock.

JoshuaMBa committed:
[scudo] Added LRU eviction policy to secondary cache.

The logic for emptying the cache now follows an LRU eviction policy.
When the cache is full on any given free operation, the oldest entry
in the cache is evicted, and the memory associated with that cache
entry is unmapped.

Finding empty cache entries is now a constant operation with the use
of a stack of available cache entries.

Through the LRU structure, the cache retrieval algorithm now only iterates
through valid entries of the cache. Furthermore, the retrieval algorithm will
first search cache entries that have not been decommitted (i.e. madvise() has
not been called on their corresponding memory chunks) to reduce the likelihood of
returning a memory chunk to the user that would induce a page fault.

cferris1000 reviewed (approved):
LGTM.
